SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -21,62 +21,113 @@ SpatialLoadBalanceEnforcer::SpatialLoadBalanceEnforcer(const PhysicalWorkerName&
 {
-	check(Op.update.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID)
-	if (StaticComponentView->HasAuthority(Op.entity_id, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+	check(Op.update.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID);
+	if (CanEnforce(Op.entity_id))
 	{
-		QueueAclAssignmentRequest(Op.entity_id);
+		MaybeQueueAclAssignmentRequest(Op.entity_id);
 	}
 }
 
-// This is called whenever this worker becomes authoritative for the ACL component on an entity.
-// It is now this worker's responsibility to make sure that the ACL reflects the current intent,
-// which may have been received before this call.
-void SpatialLoadBalanceEnforcer::AuthorityChanged(const Worker_AuthorityChangeOp& AuthOp)
+void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentAdded(const Worker_AddComponentOp& Op)
 {
-	if (AuthOp.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID &&
-		AuthOp.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+	if (!(Op.data.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID || Op.data.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID))
 	{
-		const SpatialGDK::AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(AuthOp.entity_id);
-		if (AuthorityIntentComponent == nullptr)
-		{
-			// TODO(zoning): There are still some entities being created without an authority intent component.
-			// For example, the Unreal created worker entities don't have one. Even though those won't be able to
-			// transition, we should have the intent component on them for completeness.
-		 	UE_LOG(LogSpatialLoadBalanceEnforcer, Log, TEXT("Requested authority change for entity without AuthorityIntent component. EntityId: %lld"), AuthOp.entity_id);
-			return;
-		}
+		UE_LOG(LogSpatialLoadBalanceEnforcer, Warning,
+			TEXT("Load balancer notified of add component %d for entity %lld which is not a load balancing component"),
+			Op.data.component_id, Op.entity_id);
+		return;
+	}
 
-		const PhysicalWorkerName* OwningWorkerId = VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
-		if (OwningWorkerId != nullptr &&
-			*OwningWorkerId == WorkerId &&
-			StaticComponentView->HasAuthority(AuthOp.entity_id, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
-		{
-			UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose, TEXT("No need to queue newly authoritative entity %lld because this worker is already authoritative."), AuthOp.entity_id);
-			return;
-		}
-		QueueAclAssignmentRequest(AuthOp.entity_id);
+	if (CanEnforce(Op.entity_id))
+	{
+		MaybeQueueAclAssignmentRequest(Op.entity_id);
+	}
+}
+
+void SpatialLoadBalanceEnforcer::OnLoadBalancingComponentRemoved(const Worker_RemoveComponentOp& Op)
+{
+	if (!(Op.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID || Op.component_id == SpatialConstants::ENTITY_ACL_COMPONENT_ID))
```

Comment:
Same as above

Reply:
make a check imo, receiver has a conditional for this already

Reply:
now a check